src/multivariate/jointorderstatistics.jl

```diff
@@ -0,0 +1,132 @@
+The ``i``th order statistic of a sample is the ``i``th element of the sorted sample.
+For example, the 1st order statistic is the sample minimum, while the ``n``th order
+statistic is the sample maximum.
+
+`r` must be a sorted vector of unique integers between 1 and `n`.
+
+For a single order statistic, use [`OrderStatistic`](@ref) instead.
+
+## Examples
+
+```julia
+JointOrderStatistics(Normal(), 10)           # Product(fill(Normal(), 10)) restricted to ordered vectors
+JointOrderStatistics(Cauchy(), 10, [1, 10])  # joint distribution of the extrema
+```
+"""
+struct JointOrderStatistics{D<:ContinuousUnivariateDistribution,R<:AbstractVector{Int}} <:
+       ContinuousMultivariateDistribution
+    dist::D
+    n::Int
+    r::R
+    function JointOrderStatistics(
+        dist::ContinuousUnivariateDistribution,
+        n::Int,
+        r::AbstractVector{Int}=1:n;
+        check_args::Bool=true,
+    )
+        @check_args(
+            JointOrderStatistics,
+            (n, n ≥ 1, "`n` must be a positive integer."),
+            (
+                r,
+                1 ≤ first(r) && last(r) ≤ n && issorted(r) && allunique(r),
+                "`r` must be a sorted vector of unique integers between 1 and `n`.",
+            ),
+        )
+        return new{typeof(dist),typeof(r)}(dist, n, r)
+    end
+end
+
+length(d::JointOrderStatistics) = length(d.r)
+function insupport(d::JointOrderStatistics, x::AbstractVector)
+    return length(d) == length(x) && issorted(x) && all(Base.Fix1(insupport, d.dist), x)
+end
+minimum(d::JointOrderStatistics) = fill(minimum(d.dist), length(d))
+maximum(d::JointOrderStatistics) = fill(maximum(d.dist), length(d))
+
+params(d::JointOrderStatistics) = tuple(params(d.dist)..., d.n, d.r)
+partype(d::JointOrderStatistics) = partype(d.dist)
+Base.eltype(::Type{<:JointOrderStatistics{D}}) where {D} = Base.eltype(D)
+
+function logpdf(d::JointOrderStatistics, x::AbstractVector{<:Real})
+    n = d.n
+    r = d.r
+    lp = sum(Base.Fix1(logpdf, d.dist), x)
+    T = eltype(lp)
+    lp += loggamma(T(n + 1))
+    length(r) == n && return lp
+    i = 0
+    xᵢ = oftype(float(first(x)), -Inf)
+    for (j, xⱼ) in zip(r, x)
+        lp += _marginalize_range(d.dist, n, i, j, xᵢ, xⱼ, T)
+        i = j
+        xᵢ = xⱼ
+    end
+    j = n + 1
+    xⱼ = oftype(xᵢ, Inf)
+    lp += _marginalize_range(d.dist, n, i, j, xᵢ, xⱼ, T)
+    return lp
+end
+
+# given ∏ₖf(xₖ), marginalize all xₖ for i < k < j
+function _marginalize_range(dist, n, i, j, xᵢ, xⱼ, T)
+    k = j - i - 1
+    k == 0 && return zero(T)
+    lpdiff = if i == 0
+        T(logcdf(dist, xⱼ))
+    elseif j == n + 1
+        T(logccdf(dist, xᵢ))
+    else
+        T(logdiffcdf(dist, xⱼ, xᵢ))
+    end
+    return k * lpdiff - loggamma(T(k + 1))
+end
+
+function _rand!(rng::AbstractRNG, d::JointOrderStatistics, x::AbstractVector{<:Real})
+    n = d.n
+    if n == length(d.r)  # r == 1:n
+        # direct method, slower than inversion method for large `n` and distributions with
```

Comment:
Seems we should maybe use some heuristic for choosing one of the two algorithms, depending on `n` and possibly the type of distribution?

Reply:
I'm not certain what heuristic we could use here, since which strategy is faster strongly depends on the relative speed of the quantile function and the `rand` method of the wrapped distribution. Also `n`, since the direct method has an `O(nlog(n))` sort step that for `n` large enough will make it slower than the exponential method, which is always `O(n)`, though potentially with a large constant factor.

Reply:
Hmm, no I'm not aware of such an example. Might be good enough for now.